### PR TITLE
Refactor and Fix Old Library Code

### PR DESCRIPTION
**Pull request by SimpleSandman**
## General fixes

- [x] Refactor any code that is not following the modern C# naming conventions
- [x] Fix `VORPCORE is null` issue
- [x] Add optimizations based on any glaring issues (i.e. redundant or overly-engineered code)

### Server Events
- [x] BuyHouse()
- [x] BuyRoom()
- [x] changeDoorState()
- [x] getRooms()
- [x] getHouses()

### Client Events
- [x] UpdateHousesStatus()
- [x] UpdateRoomsStatus()
- [x] SetHouseOwner()
- [x] SetDoorState()
- [x] ListHouses()
- [x] ListRooms()

### Client Ticks
- [x] UseInteriorComps()
- [x] ChangeStatus()
- [x] DoorLockeds()

### House Inventory Events
- [x] TakeFromHouse()
- [x] MoveToHouse()
- [x] UpdateInventoryHouse()

### CODE REVIEW

**Comment by SimpleSandman** [comment]
This last commit made me hate my life a little bit for how incomplete PluginManager/Manager was in relation to the child classes. I'm now able to see the blips on the map and even the draw text on doors. I've attempted to buy a house with $0 and it prompts that I don't have enough money, twice.

NOTE: I've added various error messages and warnings that didn't exist before thanks to this endeavor, especially for PlayerExtensions.cs.

There was also a minor optimization for the room lookup if the `charIdentifier` cannot be found and returns -1. We will skip the table lookup since no character should have a `charIdentifier` of -1 since that column `AUTO_INCREMENT` from 0.

Pre-alpha testing V2 to see what else is broken:
[vorp_housing.zip](https://github.com/VORPCORE/VORP-Housing/files/8569844/vorp_housing.zip)

**Comment by SimpleSandman** [comment]
Fixes:
- Removed `[Tick]` on tick methods as this was causing the duplicated actions on the client-side.
- Cleaned up some more logic and added "try/catch" statements for easier debugging.

Known Issues:
- Doors on the client-side do not "easily" open after buying a house. One has to force their way in by brute-force even though the house's "IsOpen" property is set to `true`

--------
Pre-alpha V3 to see what else is broken:
[vorp_housing.zip](https://github.com/VORPCORE/VORP-Housing/files/8599862/vorp_housing.zip)

**Comment by SimpleSandman** [comment]
Here's a video to reproduce the issue after buying a house:
https://streamable.com/metbfr

**Comment by SimpleSandman** [comment]
> Doors on the client-side do not "easily" open after buying a house. One has to force their way in via brute-force even though the house's "IsOpen" property is set to `true`

Fixed this issue by modeling after the "doorlocks" resource in the premade server zip.

**TODO**: Need to reflect changes to door locks for rooms.

----------------

Pre-alpha V4 to see what else is broken:
[vorp_housing.zip](https://github.com/VORPCORE/VORP-Housing/files/8600883/vorp_housing.zip)

**Comment by peppinoita** [comment]
image

**Comment by SimpleSandman** [comment]
The last 2 commits were dirty because of staging the forced addition to the server `.pdb` file.

NOTE: Both committed `.pdb` files are only for the "/build/vorp_housing". If or whenever we decide to move away from that folder path, we will be able to properly adhere to the `.gitignore`

**Comment by SimpleSandman** [comment]
@peppinoita would you mind testing again, but with this zip instead? I was adding some more stuff to allow easier troubleshooting on my end.

Pre-alpha V5 to see what else is broken:
[vorp_housing.zip](https://github.com/VORPCORE/VORP-Housing/files/8603210/vorp_housing.zip)

**Comment by peppinoita** [comment]
Nothing to do 

**Error client side:**

[    414297] [b1436_GTAProce]             MainThrd/ [ERROR] Client.Init.DoorLocksTickAsync()
[    414297] [b1436_GTAProce]             MainThrd/ System.NullReferenceException: Object reference not set to an instance of an object.
[    414297] [b1436_GTAProce]             MainThrd/   at VORP.Housing.Client.Init+<DoorLocksTickAsync>d__13.MoveNext () [0x0015f] in <ceffba02153044048393d1684c766e70>:0 
[    414344] [b1436_GTAProce]             MainThrd/ [ERROR] Client.Init.DoorLocksTickAsync()
[    414344] [b1436_GTAProce]             MainThrd/ System.NullReferenceException: Object reference not set to an instance of an object.
[    414344] [b1436_GTAProce]             MainThrd/   at VORP.Housing.Client.Init+<DoorLocksTickAsync>d__13.MoveNext () [0x0015f] in <ceffba02153044048393d1684c766e70>:0 

**Error server side:**
[ script:vorp_housing] [ERROR] Server.Init.MoveToHouse()
[ script:vorp_housing] Newtonsoft.Json.JsonReaderException: Unexpected character encountered while parsing value: S. Path '', line 0, position 0.
[ script:vorp_housing]   at Newtonsoft.Json.JsonTextReader.ParseValue () [0x002b3] in <d7f577efea6043649cddd7a978a2c473>:0
[ script:vorp_housing]   at Newtonsoft.Json.JsonTextReader.Read () [0x0004c] in <d7f577efea6043649cddd7a978a2c473>:0
[ script:vorp_housing]   at Newtonsoft.Json.Linq.JArray.Load (Newtonsoft.Json.JsonReader reader, Newtonsoft.Json.Linq.JsonLoadSettings settings) [0x00008] in <d7f577efea6043649cddd7a978a2c473>:0
[ script:vorp_housing]   at Newtonsoft.Json.Linq.JArray.Parse (System.String json, Newtonsoft.Json.Linq.JsonLoadSettings settings) [0x0000c] in <d7f577efea6043649cddd7a978a2c473>:0
[ script:vorp_housing]   at Newtonsoft.Json.Linq.JArray.Parse (System.String json) [0x00000] in <d7f577efea6043649cddd7a978a2c473>:0
[ script:vorp_housing]   at VORP.Housing.Server.HouseInventory.MoveToHouse (CitizenFX.Core.Player player, System.String jsonData) [0x002a8] in C:\Users\kento\source\repos\VORP-Housing-Fork\VORP-Housing\VORP.Housing.Server\HouseInventory.cs:391

**Comment by SimpleSandman** [comment]
@peppinoita Thank you for posting the messages themselves instead of a picture.

> Error client side:
[ 414297] [b1436_GTAProce] MainThrd/ [ERROR] Client.Init.DoorLocksTickAsync()
[ 414297] [b1436_GTAProce] MainThrd/ System.NullReferenceException: Object reference not set to an instance of an object.
[ 414297] [b1436_GTAProce] MainThrd/ at VORP.Housing.Client.Init+d__13.MoveNext () [0x0015f] in :0
[ 414344] [b1436_GTAProce] MainThrd/ [ERROR] Client.Init.DoorLocksTickAsync()
[ 414344] [b1436_GTAProce] MainThrd/ System.NullReferenceException: Object reference not set to an instance of an object.
[ 414344] [b1436_GTAProce] MainThrd/ at VORP.Housing.Client.Init+d__13.MoveNext () [0x0015f] in :0

It's odd that you're still getting the same message whenever you have the .pdb files. Are you sure you've updated the `config.json`? Because I did add a new property called, "DoorHashes", that allow easier look ups on doors. Otherwise, I'll have to add more check for anything null and spit out an error message explaining why.

> Error server side:
[ script:vorp_housing] [ERROR] Server.Init.MoveToHouse()
[ script:vorp_housing] Newtonsoft.Json.JsonReaderException: Unexpected character encountered while parsing value: S. Path '', line 0, position 0.
[ script:vorp_housing] at Newtonsoft.Json.JsonTextReader.ParseValue () [0x002b3] in :0
[ script:vorp_housing] at Newtonsoft.Json.JsonTextReader.Read () [0x0004c] in :0
[ script:vorp_housing] at Newtonsoft.Json.Linq.JArray.Load (Newtonsoft.Json.JsonReader reader, Newtonsoft.Json.Linq.JsonLoadSettings settings) [0x00008] in :0
[ script:vorp_housing] at Newtonsoft.Json.Linq.JArray.Parse (System.String json, Newtonsoft.Json.Linq.JsonLoadSettings settings) [0x0000c] in :0
[ script:vorp_housing] at Newtonsoft.Json.Linq.JArray.Parse (System.String json) [0x00000] in :0
[ script:vorp_housing] at VORP.Housing.Server.HouseInventory.MoveToHouse (CitizenFX.Core.Player player, System.String jsonData) [0x002a8] in C:\Users\kento\source\repos\VORP-Housing-Fork\VORP-Housing\VORP.Housing.Server\HouseInventory.cs:391

I haven't touched anything on the `HouseInventory` yet. So I'm not surprised there's an issue with the parsing. But I'm happy it got as far as it did since I can see the line number on this error at 391.

**Comment by peppinoita** [comment]
Give me a sec, I didn't user the default config, I used one of another language (inside the download) sorry let me check again

**Comment by peppinoita** [comment]
Now I do not have any error on client side but the door doesn't open. Do I need additional script to manage doors?

On server side same errors with the inventory

**Comment by peppinoita** [comment]
Ok I'm really sorry, I checked all the scripts in my test server and I found an old doorlock without icons to be active.
Without it the doorlock works really good.
There is only the inventory error now and I hope this isn't my fault again. I will check.
I'm using latest inventory lua

**Comment by SimpleSandman** [comment]
Fixed door locks for rooms and inventory management for this script.

Beta testing:
[vorp_housing.zip](https://github.com/VORPCORE/VORP-Housing/files/8627787/vorp_housing.zip)

**Comment by SimpleSandman** [comment]
@peppinoita I've also updated the other `config.json` files so you can use your preferred language

**Comment by EagLeTaLoN713** [comment]
I will drop it in server tomorrow and play with it :) 
Thanks for all the work man!

**Comment by EagLeTaLoN713** [comment]
I have installed and added the custom house and still unable to get it to recognise inventory in the specified area.

**Comment by EagLeTaLoN713** [comment]
{
     "Id": 14701,
      "DoorHashes": [ 2472180395, 2188892390 ],
      "DoorModelHashes": [ 1411208969 ],
     "InteriorName": "shb_p_ammo01",
      "Name": "Dodge Ranch",
     "Price":1,
     "DoorsStatus": [ 1779.35, 460.92, 113.72 ],
      "Doors": [
        [ 1779.5008544922,461.01312255859,112.03468322754, 105.00004577637 ],
        [ 1789.1323242188,463.61395263672,112.03468322754, 104.93343353271 ]
      ],
      "Inventory": [ 1781.971, 466.797, 112.0378, 1.5 ],
      "MaxWeight": 70
    },

**Comment by EagLeTaLoN713** [comment]
Also after purchasing a house, you do not get the call to open or close the door

**Comment by SimpleSandman** [comment]
Just as a note for myself:
- [Dodge House door hash](https://github.com/femga/rdr3_discoveries/blob/master/doorHashes/doorhashes.lua#L434)
- Map location

image

- Picture of house

image

**Comment by SimpleSandman** [comment]
So @EagLeTaLoN713 I can say that the house does show up after a server restart, but you have to clear the cache in order to get it to show up. That's something I unfortunately cannot fix since that's a CFX issue.

> Also after purchasing a house, you do not get the call to open or close the door

This was my bad for not updating the client-side with the house having the owner's `charIdentifier` (your character's ID). This ensures only your character can own a house/room.

> "Inventory": [ 1781.971, 466.797, 112.0378, 1.5 ]

I would have to look into why the inventory itself isn't popping up for this house given the coordinates you've provided. The picture below shows where I would go for the inventory in this house.

image

Does the inventory show up with the old script? If so, then I'll have to do some digging into why this isn't showing up. Otherwise, I'm curious if the inventory coordinates need to be moved somewhere else.

----------------------

Here's the newest zip for you to test:
[vorp_housing.zip](https://github.com/VORPCORE/VORP-Housing/files/8633182/vorp_housing.zip)

**Comment by EagLeTaLoN713** [comment]
That’s the spot in the house and inventory has never worked. Thank you

**Comment by peppinoita** [comment]
works pretty well but doesn't change the icons from free to purchased

**Comment by SimpleSandman** [comment]
> works pretty well but doesn't change the icons from free to purchased

@peppinoita, are you talking about the map icons? If so, I will have to take a look into how that could be arranged, but it wouldn't be a high priority fix. If anything, I wouldn't mind you creating a GitHub issue after the release/merge of this PR. This way, if I don't get to it, someone else can.

**Comment by peppinoita** [comment]
ok, thanks for the effort you put in this rework.

**Comment by SimpleSandman** [comment]
@peppinoita No problem! Glad I can make this maintainable and usable again!

**Comment by EagLeTaLoN713** [comment]
Btw 14701 was just a made up number and the interior name was just a placeholder as there is no interior for it.

**Comment by SimpleSandman** [comment]
> Btw 14701 was just a made up number and the interior name was just a placeholder as there is no interior for it.

The timing of you mentioning this and me committing my inventory fix...are you watching me?????

------------------

BTW, the newest of new zips for you to test:
[vorp_housing.zip](https://github.com/VORPCORE/VORP-Housing/files/8636573/vorp_housing.zip)

**Comment by EagLeTaLoN713** [comment]
Maybe 🤔 :)

**Comment by EagLeTaLoN713** [comment]
I’ll install in the AM and test. You a smart mutha fuka!

**Comment by SimpleSandman** [comment]
> Installed and tested inventory on Dodge Ranch, and I noticed Inventory still doesn't work.

I forgot to mention that I did move it to the fireplace in the main room @EagLeTaLoN713

**Comment by EagLeTaLoN713** [comment]
Well that might be why Haha.  Ill go check

**Comment by EagLeTaLoN713** [comment]
> > Installed and tested inventory on Dodge Ranch, and I noticed Inventory still doesn't work.
> 
> I forgot to mention that I did move it to the fireplace in the main room @EagLeTaLoN713

I can move it though right?

**Comment by SimpleSandman** [comment]
> I can move it though right?

Oh of course! Just reassign the inventory's coordinates. I moved it to the fireplace because from a RP POV, I would grab something from the fireplace like a scene from the RDR2 story.

**Comment by EagLeTaLoN713** [comment]
> > I can move it though right?
> 
> Oh of course! Just reassign the inventory's coordinates. I moved it to the fireplace because from a RP POV, I would grab something from the fireplace like a scene from the RDR2 story.

ahh never played story, but in my head I would think a cabinet haha.  Also, I think there is an issue. While I get the prompt for the inventory, I am unable drag and drop anything in it as it shows both inventories empty sometimes. Also, restarting the script removes ownership of the house and you have to rebuy it.

**Comment by EagLeTaLoN713** [comment]
Okay, after leaving the area of the FP, you can see your inventory when you enter the storage, but you can't move anything to it.

**Comment by SimpleSandman** [comment]
> ahh never played story, but in my head I would think a cabinet haha.

Lol you're good. On my side, it's an empty house and it looked like you were trying to grab something from a secret compartment in the wall

> While I get the prompt for the inventory, I am unable drag and drop anything in it as it shows both inventories empty sometimes. Also, restarting the script removes ownership of the house and you have to rebuy it.

I had to personally reassign the ID inside the `housing` table in order to properly get my inventory since the "interior entity ID" is the main factor. The old ID you sent me was the reason why I wasn't able to get the prompt in the first place. You could say it was "bad data" that needed to be cleared from the database for testing.

> Okay, after leaving the area of the FP, you can see your inventory when you enter the storage, but you can't move anything to it.

I'm using the latest C# inventory and I'm able to move stuff in and out just fine. I haven't tested the Lua version because I know it's still new in comparison.

**Comment by EagLeTaLoN713** [comment]
Ah, we are all using the LUA as its the newest update.  Any reason the ownership is reset when the script is reset?  Going to restart server just to be sure its not in that as an issue.

**Comment by EagLeTaLoN713** [comment]
Okay after server restart I had to buy the house again and it worked as intended.

**Comment by SimpleSandman** [comment]
> Ah, we are all using the LUA as its the newest update.

After this is merged, please report that to the LUA repo as it works in C#. I'm only a C# coder so I wouldn't have any idea as to why it would be breaking over there.

> Any reason the ownership is reset when the script is reset?
> Okay after server restart I had to buy the house again and it worked as intended.

I do a server restart every time I push any kind of code and I haven't experienced this issue. It must be the "old ID" being in the `housing` table that caused this issue. You can also be experiencing a caching issue from CFX.

**Comment by EagLeTaLoN713** [comment]
Any chance we can voice Sandman So i can explain something to you?

**Comment by SimpleSandman** [comment]
> Any chance we can voice Sandman So i can explain something to you?

Unfortunately, I just started my work day as I'm UTC-5 (10:17 AM as of this message) so I won't be able to until after 5 PM UTC-5. We can DM in Discord if you have more to talk about.

**Comment by SimpleSandman** [comment]
Fixed the check for "if a table exists" by dynamically looking for the current database name using the `DATABASE()` sql function. This will help people that have different database names than what's in the premade.

I want this to be my last ZIP before release, but if people find more issues, then keep the reports coming!
[vorp_housing.zip](https://github.com/VORPCORE/VORP-Housing/files/8646146/vorp_housing.zip)

**Comment by Sina30** [comment]
You get the error in the console if a player has a house and comes to the server

[ script:vorp_housing] [DEBUG] [{"id":54531,"identifier":"steam:0000000000","charidentifier":635,"inventory":null,"furniture":"{}","open":true}]
[ script:vorp_housing] [ERROR] Server.Init.GetHousesAsync()
[ script:vorp_housing] System.Collections.Generic.KeyNotFoundException: The given key '54531' was not present in the dictionary.
[ script:vorp_housing]   at System.Collections.Generic.Dictionary`2[TKey,TValue].get_Item (TKey key) [0x0001e] in <fbc4ec45371543bfba3678ebb82caf6d>:0 
[ script:vorp_housing]   at VORP.Housing.Server.Init.GetHousesAsync (System.Int32 source) [0x00af3] in C:\Users\kento\source\repos\VORP-Housing-Fork\VORP-Housing\VORP.Housing.Server\Init.cs:156

**Comment by EagLeTaLoN713** [comment]
I would completely wipe your housing db. and rerun the SQL file. Ive put it here.   And have everyone "rebuy" homes as for some reason the new script and sql will fight the old entries.  This worked for me

[housing.zip](https://github.com/VORPCORE/VORP-Housing/files/8650180/housing.zip)

**Comment by EagLeTaLoN713** [comment]
oh and save any inventory they have on a document so you can reenter the information back into the inventory block after wipe so they don't loose their items

File: VORP-Housing/VORP.Housing.Server/Init.cs [hunk 1]
```diff
@@ -0,0 +1,315 @@
...
+using System.Threading.Tasks;
+using VORP.Housing.Server.Extensions;
+using VORP.Housing.Server.Scripts;
+using VORP.Housing.Shared;
+using VORP.Housing.Shared.Models;
+using VORP.Housing.Shared.Models.Json;
+
+namespace VORP.Housing.Server
+{
+    public class Init : Manager
```
Comment (Local9) [inline VORP-Housing/VORP.Housing.Server/Init.cs:13]: Would name the class more around what it is doing, I understand its an "Init" but an initiation of what?

Comment (SimpleSandman) [reply VORP-Housing/VORP.Housing.Server/Init.cs:13]: To be real with you, I haven't figured out a decent name to replace the old `vorphousing_cl_init` or `vorphousing_sv_init` classes

Comment (SimpleSandman) [reply VORP-Housing/VORP.Housing.Server/Init.cs:13]: I guess I could call it `InitHousing`, but I'm open to suggestions

File: VORP-Housing/VORP.Housing.Server/Init.cs [hunk 2]
```diff
@@ -0,0 +1,315 @@
...
+        public static Dictionary<uint, House> HousesDb = new Dictionary<uint, House>();
+        public static Dictionary<int, Room> RoomsDb = new Dictionary<int, Room>();
+        
+        public void Initialize()
+        {
+            AddEvent("vorp_housing:BuyHouse", new Action<Player, uint, double>(BuyHouseAsync));
+            AddEvent("vorp_housing:BuyRoom", new Action<Player, int, double>(BuyRoomAsync));
+            AddEvent("vorp_housing:changeDoorState", new Action<uint, bool>(ChangeDoorState));
+            AddEvent("vorp_housing:getRooms", new Action<int>(GetRoomsAsync));
+            AddEvent("vorp_housing:getHouses", new Action<int>(GetHousesAsync));
```
Comment (Local9) [inline VORP-Housing/VORP.Housing.Server/Init.cs:26]: Event naming I believe we should discuss if it should either have "EventHandler" in the method name, or have "On" as a prefix for an event handler method. This is for all C# projects, not just this one.

I've been using "On" prefixing, but a discussion is best having.

Comment (SimpleSandman) [reply VORP-Housing/VORP.Housing.Server/Init.cs:26]: I wouldn't mind using the "On" prefix. I haven't written as many event handlers as I've should.

Comment (SimpleSandman) [reply VORP-Housing/VORP.Housing.Server/Init.cs:26]: Never mind, based on [the official docs from Microsoft](https://docs.microsoft.com/en-us/dynamicsax-2012/developer/naming-conventions-delegates-and-event-handlers), an "EventHandler" suffix is the suggested way

**Review by Local9 — APPROVED** [review]
Only comments added for discussion or clarification.

**Comment by SimpleSandman** [comment]
Awesome! Let me change those two things you mentioned real quick before merging

File: VORP-Housing/VORP.Housing.Server/Init.cs (hunk 1 above)
Comment (Local9) [reply VORP-Housing/VORP.Housing.Server/Init.cs:13]: Does look that the class is only dealing with house events, the resource naming makes it more confusing.

File: VORP-Housing/VORP.Housing.Server/Init.cs (hunk 2 above)
Comment (Local9) [reply VORP-Housing/VORP.Housing.Server/Init.cs:26]: That is the X++ documentation, its fine to use EventHandler, the C# naming conventions isn't clear about it either. The "On" handling is from ASP .NET event naming e.g. `btnSubmit_OnClick`

Comment (Local9) [reply VORP-Housing/VORP.Housing.Server/Init.cs:26]: Could be more, we decide what we're happier with; https://stackoverflow.com/questions/724085/events-naming-convention-and-style

**Comment by SimpleSandman** [comment]
This is ready to be merged.

Take note of @EagLeTaLoN713's recommendations to ensure an easy transition to this reworked resource:

> I would completely wipe your housing db. and rerun the SQL file and have everyone "rebuy" homes as for some reason the new script and sql will fight the old entries. This worked for me.
> Save any inventory they have on a document so you can reenter the information back into the inventory block after wipe so they don't loose their items

**Comment by Local9** [comment]
Were there changes to the SQL or is it a JSON issue in the DB?

**Comment by Local9** [comment]
I gather you had some example data to base the classes on? and its just a casing issue between one and the other?

**Comment by SimpleSandman** [comment]
> I gather you had some example data to base the classes on? and its just a casing issue between one and the other?

I did not have any example table data to verify the structural integrity, but I didn't mess with any of the SQL queries inside the code that could've altered the structure. A lot of my changes were based on inference of the existing code. I did my best to not touch the actual structure of said data other than the casing to the best of my knowledge. I can assure you I didn't touch the initial SQL script itself for creating the tables.

This entire PR was supposed to be to keep as much of the functionality as close as I could to the original while restructuring how the actual code was laid out for future maintainability.

**Comment by Local9** [comment]
Thats fine, I gather the JSON outputs you've got from this are using the `JsonProperty` attribute correctly? as I've just always used the `DataMember` attribute.

**Comment by SimpleSandman** [comment]
> Thats fine, I gather the JSON outputs you've got from this are using the `JsonProperty` attribute correctly? as I've just always used the `DataMember` attribute.

Correct! So when I was looking up the differences for `DataMember` versus `JsonProperty` is actually a version difference in the .NET framework historically. `DataMember` is the Microsoft predecessor to the Newtonsoft `JsonProperty`. Now this next bit doesn't apply to us because of .NET 5, but what succeeds Newtonsoft is Microsoft's in-house version in the `System.Text.Json` library.

**Comment by SimpleSandman** [comment]
Since I haven't seen any movement in this PR, I'll add my reasoning to use `JsonProperty`. 

The reason why I decided to use that attribute versus `DataMember` is because in the event we are able to work on newer .NET libraries that utilize `System.Text.Json` we can have an easy migration from Newtonsoft. Newtonsoft itself hasn't been updated since March 2021 (last I've checked). The migration process itself is well documented on Microsoft's website. If nothing else, Newtonsoft has loads of documentation that is FAR more recent than `DataMember` while still being extensive and well-tested.

**Pull request merged**